In [2]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 1: No Time for a Taxicab ---</h2><p>Santa's sleigh uses a <span title="An atomic clock is too inaccurate; he might end up in a wall!">very high-precision clock</span> to guide its movements, and the clock's oscillator is regulated by stars. Unfortunately, the stars have been stolen... by the Easter Bunny.  To save Christmas, Santa needs you to retrieve all <em class="star">fifty stars</em> by December 25th.</p>
<p>Collect stars by solving puzzles.  Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first.  Each puzzle grants <em class="star">one star</em>. Good luck!</p>
<p>You're airdropped near <em>Easter Bunny Headquarters</em> in a city somewhere.  "Near", unfortunately, is as close as you can get - the instructions on the Easter Bunny Recruiting Document the Elves intercepted start here, and nobody had time to work them out further.</p>
<p>The Document indicates that you should start at the given coordinates (where you just landed) and face North.  Then, follow the provided sequence: either turn left (<code>L</code>) or right (<code>R</code>) 90 degrees, then walk forward the given number of blocks, ending at a new intersection.</p>
<p>There's no time to follow such ridiculous instructions on foot, though, so you take a moment and work out the destination.  Given that you can only walk on the <a href="https://en.wikipedia.org/wiki/Taxicab_geometry">street grid of the city</a>, how far is the shortest path to the destination?</p>
<p>For example:</p>
<ul>
<li>Following <code>R2, L3</code> leaves you <code>2</code> blocks East and <code>3</code> blocks North, or <code>5</code> blocks away.</li>
<li><code>R2, R2, R2</code> leaves you <code>2</code> blocks due South of your starting position, which is <code>2</code> blocks away.</li>
<li><code>R5, L5, R5, R3</code> leaves you <code>12</code> blocks away.</li>
</ul>
<p><em>How many blocks away</em> is Easter Bunny HQ?</p>
</article>


In [28]:
tests = [
    {
        "name": "Example 1",
        "path": "R2, L3",
        "expected": 5,
    },
    {
        "name": "Example 2",
        "path": "R2, R2, R2",
        "expected": 2,
    },
    {
        "name": "Example 3",
        "path": "R5, L5, R5, R3",
        "expected": 12,
    },
    {
        "name": "Example 4",
        "path": "R500, L1200",
        "expected": 1700,
    },
]


def distance(path: str) -> int:
    direction = [-1, 0]
    position = [0, 0]
    for step in re.split(r",\s+", path):
        turn, blocks = step[0], int(step[1:])
        if direction == [-1, 0]:  # Nort
            if turn == "R":
                direction = [0, 1]  # East
            else:  # turn == L
                direction = [0, -1]  # West
        elif direction == [0, 1]:  # East
            if turn == "R":
                direction = [1, 0]  # South
            else:  # turn == L
                direction = [-1, 0]  # North
        elif direction == [1, 0]:  # South
            if turn == "R":
                direction = [0, -1]  # West
            else:  # turn == L
                direction = [0, 1]  # East
        else:  # [0, -1] West
            if turn == "R":
                direction = [-1, 0]  # North
            else:  # turn == L
                direction = [1, 0]  # South

        position = [p + blocks * d for p, d in zip(position, direction)]
    return sum(abs(p) for p in position)


run_tests_params(distance, tests)


Test Example 1 passed, for distance.
Test Example 2 passed, for distance.
Test Example 3 passed, for distance.
Test Example 4 passed, for distance.
Success


In [29]:
with open("../input/day1.txt") as f:
    print(distance(f.read()))

301


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>301</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Then, you notice the instructions continue on the back of the Recruiting Document.  Easter Bunny HQ is actually at the first location you visit twice.</p>
<p>For example, if your instructions are <code>R8, R4, R4, R8</code>, the first location you visit twice is <code>4</code> blocks away, due East.</p>
<p>How many blocks away is the <em>first location you visit twice</em>?</p>
</article>

</main>


In [34]:
def distance_first_loction_visited_twice(path: str) -> int:
    direction = [-1, 0]
    position = (0, 0)
    visited = {position}
    for step in re.split(r",\s+", path):
        turn, blocks = step[0], int(step[1:])
        if direction == [-1, 0]:  # Nort
            if turn == "R":
                direction = [0, 1]  # East
            else:  # turn == L
                direction = [0, -1]  # West
        elif direction == [0, 1]:  # East
            if turn == "R":
                direction = [1, 0]  # South
            else:  # turn == L
                direction = [-1, 0]  # North
        elif direction == [1, 0]:  # South
            if turn == "R":
                direction = [0, -1]  # West
            else:  # turn == L
                direction = [0, 1]  # East
        else:  # [0, -1] West
            if turn == "R":
                direction = [-1, 0]  # North
            else:  # turn == L
                direction = [1, 0]  # South

        for _ in range(1, blocks + 1):
            position = tuple(p + d for p, d in zip(position, direction))
            if position in visited:
                return sum(abs(p) for p in position)
            visited.add(position)

    return -1


distance_first_loction_visited_twice("R8, R4, R4, R8")

4

In [35]:
with open("../input/day1.txt") as f:
    print(distance_first_loction_visited_twice(f.read()))

130


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>130</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2016">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="1/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22No+Time+for+a+Taxicab%22+%2D+Day+1+%2D+Advent+of+Code+2016&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2016%2Fday%2F1&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22No+Time+for+a+Taxicab%22+%2D+Day+1+%2D+Advent+of+Code+2016+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2016%2Fday%2F1';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
